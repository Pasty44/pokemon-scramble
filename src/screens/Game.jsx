import { useCallback, useEffect, useRef, useState } from 'react';
import { shuffleArray, shuffleString } from 'scripts/util';

const INPUT_COLORS = {
    NEUTRAL: 'black',
    SUCCESS: '#00BB00',
    FAIL: '#DD0000'
};

export default function Game({ pokemonList, reset }) {
    const inputRef = useRef(null);
    const inputColor = useRef(INPUT_COLORS.NEUTRAL);

    // This is so we can clear timer on component unmounting to prevent crashes/memory leaks
    const newPokemonTimeout = useRef(null);
    useEffect(() => {
        return () => {
            if (!!newPokemonTimeout.current) clearTimeout(newPokemonTimeout.current);
        }
    }, []);

    const [currentStreak, setCurrentStreak] = useState(0);
    const [longestStreak, setLongestStreak] = useState(0);

    // Create a shuffled copy of the list prop to iterate through
    const currentPokemonList = useRef(shuffleArray([...pokemonList]));
    const currentIndex = useRef(0);
    const getCurrentPokemonObj = () => ({
        ...currentPokemonList.current[currentIndex.current],
        name: currentPokemonList.current[currentIndex.current].name.toUpperCase(),
        shuffledName: shuffleString(currentPokemonList.current[currentIndex.current].name).toUpperCase()
    });

    const [canType, setCanType] = useState(true);
    const [userGuess, setUserGuess] = useState('');
    const [currentPokemon, setCurrentPokemon] = useState(getCurrentPokemonObj);
    
    const selectNewPokemon = () => {
        // When we reach the end, shuffle and start over
        if (currentIndex.current === currentPokemonList.current.length - 1) {
            currentPokemonList.current = shuffleArray(pokemonList);
            currentIndex.current = 0;
        }
        else {
            currentIndex.current += 1;
        }
        
        setUserGuess('');
        
        inputRef.current.focus();
        
        setCurrentPokemon(getCurrentPokemonObj());
    };

    // Select new pokemon after a delay
    const startNewPokemonTimer = useCallback(() => {
        newPokemonTimeout.current = setTimeout(() => {
            inputColor.current = INPUT_COLORS.NEUTRAL;
            selectNewPokemon();
            setCanType(true);
        }, 1600);
    }, []);
    
    const onType = (updatedGuess, cursorPosition) => {
        if (!canType) return;
        
        setUserGuess(updatedGuess);

        // Prevent weird React behaviour, where putting the cursor in the middle of the guess and
        // typing a character would jump the cursor to the end of the string
        setTimeout(() => {
            inputRef.current.selectionStart = cursorPosition;
            inputRef.current.selectionEnd = cursorPosition;
        }, 0);
        
        if (updatedGuess === currentPokemon.name) {
            inputColor.current = INPUT_COLORS.SUCCESS;
            setCanType(false);
            setCurrentStreak(curr => curr + 1);

            if (currentStreak + 1 > longestStreak) {
                setLongestStreak(curr => curr + 1);
            }
            
            startNewPokemonTimer();
        }
    };

    const skip = useCallback(() => {
        inputColor.current = INPUT_COLORS.FAIL;
        setCanType(false);
        setCurrentStreak(0);
        setUserGuess(currentPokemon.name);

        startNewPokemonTimer();
    }, [currentPokemon, startNewPokemonTimer]);

    return (
        <>
            <div id='spriteContainer'>
                {canType && <div id='spriteMask'>?</div>}
                <img src={currentPokemon.url} height='100%' width='100%' alt='pokemon sprite' />
            </div>

            <div className='my-3'>
                {currentPokemon.shuffledName}
            </div>

            <div id='streakContainer'>
                <div className='streak'>
                    <div className='streakType'>
                        Current streak
                    </div>
                    <div>: {currentStreak}</div>
                </div>
                <div className='streak'>
                    <div className='streakType'>
                        Longest streak
                    </div>
                    <div>: {longestStreak}</div>
                </div>
            </div>

            {currentPokemon.name.indexOf(' ') >-1 && (
                <div id='spaceWarning'>* This name contains a space</div>
            )}

            <input
                id='userGuess'
                className='my-1'
                value={userGuess}
                onChange={(e) => onType(e.target.value.toUpperCase(), e.target.selectionStart)}
                ref={inputRef}
                autoFocus
                autoComplete='off'
                style={{ color: inputColor.current }}
            />

            <button id='skipBtn' className='my-1 noSelect' onClick={skip}>
                I DON'T KNOW
            </button>

            <button id='resetBtn' className='my-1 noSelect' onClick={reset}>
                CHANGE GENS
            </button>
        </>
    );
}
