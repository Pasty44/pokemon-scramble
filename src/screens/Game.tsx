import { useCallback, useEffect, useRef, useState } from "react";
import usePokemonQuiz from "hooks/usePokemonQuiz";
import { TPokemon } from "pokemon/types";

const INPUT_COLORS = {
  NEUTRAL: "#000000",
  SUCCESS: "#00BB00",
  FAIL: "#DD0000",
};

export interface IGameProps {
  pokemonList: Array<TPokemon>;
  reset: () => void;
}

export default function Game({ pokemonList, reset }: IGameProps) {
  const quiz = usePokemonQuiz(pokemonList);
  const inputRef = useRef<any>(null);
  const inputColor = useRef(INPUT_COLORS.NEUTRAL);

  const newPokemonTimeout = useRef<ReturnType<typeof setTimeout> | null>(null);
  // This is so we can clear timer on component unmounting to prevent crashes/memory leaks
  useEffect(() => {
    return () => {
      if (!!newPokemonTimeout.current) clearTimeout(newPokemonTimeout.current);
    };
  }, []);

  const [canType, setCanType] = useState(true);
  const [userGuess, setUserGuess] = useState("");

  // Select new pokemon after a delay
  const startNewPokemonTimer = useCallback(() => {
    const selectNewPokemon = () => {
      setUserGuess("");
      inputRef.current.focus();
      quiz.goNext();
      inputColor.current = INPUT_COLORS.NEUTRAL;
      setCanType(true);
    };

    newPokemonTimeout.current = setTimeout(selectNewPokemon, 2000);
  }, [quiz]);

  const onType = (updatedGuess: string, cursorPosition: number | null) => {
    if (!canType) return;

    setUserGuess(updatedGuess);

    // Prevent weird React behaviour, where putting the cursor in the middle of the guess and
    // typing a character would jump the cursor to the end of the string
    setTimeout(() => {
      inputRef.current.selectionStart = cursorPosition;
      inputRef.current.selectionEnd = cursorPosition;
    }, 0);

    if (updatedGuess === quiz.currentPokemon.name) {
      inputColor.current = INPUT_COLORS.SUCCESS;
      setCanType(false);
      quiz.incrementCurrentStreak();
      startNewPokemonTimer();
    }
  };

  const skip = useCallback(() => {
    inputColor.current = INPUT_COLORS.FAIL;
    setCanType(false);
    quiz.resetCurrentStreak();
    setUserGuess(quiz.currentPokemon.name);
    startNewPokemonTimer();
  }, [quiz, startNewPokemonTimer]);

  return (
    <>
      <div id="spriteContainer">
        {canType && <div id="spriteMask">?</div>}
        <img
          src={quiz.currentPokemon.url}
          height="100%"
          width="100%"
          alt="pokemon sprite"
        />
      </div>

      <div className="my-3">{quiz.currentPokemon.shuffledName}</div>

      <div id="streakContainer">
        <div className="streak">
          <div className="streakType">Current streak</div>
          <div>: {quiz.currentStreak}</div>
        </div>
        <div className="streak">
          <div className="streakType">Longest streak</div>
          <div>: {quiz.longestStreak}</div>
        </div>
      </div>

      {quiz.currentPokemon.name.indexOf(" ") > -1 && (
        <div id="spaceWarning">* This name contains a space</div>
      )}

      <input
        id="userGuess"
        className="my-1"
        value={userGuess}
        onChange={(e) =>
          onType(e.target.value.toUpperCase(), e.target.selectionStart)
        }
        ref={inputRef}
        autoFocus
        autoComplete="off"
        style={{ color: inputColor.current }}
      />

      <button id="skipBtn" className="my-1 noSelect" onClick={skip}>
        I DON'T KNOW
      </button>

      <button id="resetBtn" className="my-1 noSelect" onClick={reset}>
        CHANGE GENS
      </button>
    </>
  );
}
