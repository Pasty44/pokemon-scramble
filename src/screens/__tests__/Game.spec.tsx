import renderer from "react-test-renderer";
import Game from "screens/Game";

jest.mock("scripts/util", () => ({
  shuffleArray: (input: any) => input,
  shuffleString: (input: any) => input,
}));

const samplePokemon = {
  name: "bulbasaur",
  url: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png",
  pokedexId: 1,
};

describe("Game", () => {
  // TODO: test case with empty pokemon list - should probably show error

  it("should render game page with sample pokemon", () => {
    const component = renderer.create(
      <Game pokemonList={[samplePokemon]} reset={() => null} />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
