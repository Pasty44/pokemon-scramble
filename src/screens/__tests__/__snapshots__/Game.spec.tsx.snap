// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game should render game page with sample pokemon 1`] = `
Array [
  <div
    id="spriteContainer"
  >
    <div
      id="spriteMask"
    >
      ?
    </div>
    <img
      alt="pokemon sprite"
      height="100%"
      src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png"
      width="100%"
    />
  </div>,
  <div
    className="my-3"
  >
    BULBASAUR
  </div>,
  <div
    id="streakContainer"
  >
    <div
      className="streak"
    >
      <div
        className="streakType"
      >
        Current streak
      </div>
      <div>
        : 
        0
      </div>
    </div>
    <div
      className="streak"
    >
      <div
        className="streakType"
      >
        Longest streak
      </div>
      <div>
        : 
        0
      </div>
    </div>
  </div>,
  <input
    autoComplete="off"
    autoFocus={true}
    className="my-1"
    id="userGuess"
    onChange={[Function]}
    style={
      Object {
        "color": "#000000",
      }
    }
    value=""
  />,
  <button
    className="my-1 noSelect"
    id="skipBtn"
    onClick={[Function]}
  >
    I DON'T KNOW
  </button>,
  <button
    className="my-1 noSelect"
    id="resetBtn"
    onClick={[Function]}
  >
    CHANGE GENS
  </button>,
]
`;
